.github/workflows/test_and_deploy.yml

```diff
@@ -7,46 +7,49 @@ on:
 
 jobs:
   test:
+   #####----- RUN FULL TEST SUITE WITHOUT COVERAGE
+   # Test through actual installations for correct pre-deployment end-to-end
+   # tests.
    strategy:
      matrix:
-       os: [ubuntu-latest, macos-latest]
+       os: [ubuntu-20.04, ubuntu-22.04, macos-12, macos-13, macos-latest]
+       python-version: ["3.9", "3.10", "3.11", "3.12"]
    runs-on: ${{ matrix.os }}
+   env:
+     CLONE_PATH: ${{ github.workspace }}
    steps:
      - uses: actions/checkout@v4 
-     - name: Setup python 3.10
-       uses: actions/setup-python@v4
+     - name: Setup Python ${{ matrix.python-version }}
+       uses: actions/setup-python@v5
        with:
-         python-version: '3.10'
-     - name: Install dependencies
+         python-version: ${{ matrix.python-version }}
+     - name: Install OpenMPI
        run: |
-         python -m pip install --upgrade pip
-         pip install -e .
-     - name: Run pytests (Linux)
-       if: runner.os == 'Linux'
+         if   [ "${{ runner.os }}" == "Linux" ]; then
+            sudo apt-get update
+            sudo apt-get -y install openmpi-bin
+         elif [ "${{ runner.os }}" == "macOS" ]; then
+            /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+            brew install open-mpi
+         fi
+     - name: Setup base Python environment
        run: |
-         wget https://download.open-mpi.org/release/open-mpi/v4.0/openmpi-4.0.2.tar.gz
-         tar -xvf ./openmpi-4.0.2.tar.gz
-         ./openmpi-4.0.2/configure --prefix="/home/${USER}/.openmpi"
-         make -j
-         make install
-         export PATH=${PATH}:/home/${USER}/.openmpi/bin/
-         export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/home/${USER}/.openmpi/lib/
-         pytest --cov=Taweret test/
-     - name: Run pytest (MacOS)
-       if: runner.os == 'macOS'
+         $CLONE_PATH/.github/workflows/setup_base_python.sh ${{ runner.os }}
+     - name: Run pytests without coverage
        run: |
-         /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
-         brew install open-mpi
-         pip show openbtmixing
-         pytest --cov=Taweret test/
-     - name: Upload coverage reports to Codecov
-       uses: codecov/codecov-action@v4.0.1
-       with:
-         token: ${{ secrets.KEVIN_CODECOV_TOKEN }}
-         slug: bandframework/Taweret
+         # TODO: Is this check an extra layer of sanity checks on top of the PR
```

Comment:
Yes, this is really the last line of defense, in case all the other tests didn't work

Reply:
Sounds like a good idea to me!